presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchClient.java

```diff
@@ -0,0 +1,264 @@
+
+public class ElasticsearchClient
+{
+    private static final Logger log = Logger.get(ElasticsearchClient.class);
+
+    private final ElasticsearchConnectorConfig config;
+    private final Map<SchemaTableName, ElasticsearchTableDescription> tableDescriptions;
+
+    private Map<String, Client> clients;
+
+    @Inject
+    public ElasticsearchClient(ElasticsearchConnectorConfig elasticsearchConfig, Map<SchemaTableName, ElasticsearchTableDescription> descriptions)
+            throws IOException
+    {
+        config = elasticsearchConfig;
+        tableDescriptions = descriptions;
+        createClients();
+    }
+
+    public ElasticsearchConnectorConfig getConfig()
+    {
+        return config;
+    }
+
+    public ImmutableMap<String, Client> getClients()
+    {
+        return ImmutableMap.copyOf(clients);
+    }
+
+    public List<String> listSchemas()
+    {
+        ImmutableSet.Builder<String> builder = ImmutableSet.builder();
+        for (SchemaTableName tableName : tableDescriptions.keySet()) {
+            builder.add(tableName.getSchemaName());
+        }
+        return ImmutableList.copyOf(builder.build());
+    }
+
+    public List<SchemaTableName> listTables(String schemaNameOrNull)
+    {
+        ImmutableList.Builder<SchemaTableName> builder = ImmutableList.builder();
+        for (SchemaTableName tableName : tableDescriptions.keySet()) {
+            if (schemaNameOrNull == null || tableName.getSchemaName().equals(schemaNameOrNull)) {
+                builder.add(tableName);
+            }
+        }
+
+        return builder.build();
+    }
+
+    public ElasticsearchTableDescription getTable(String schemaName, String tableName)
+    {
+        requireNonNull(schemaName, "schemaName is null");
+        requireNonNull(tableName, "tableName is null");
+        try {
+            for (Map.Entry<SchemaTableName, ElasticsearchTableDescription> entry : tableDescriptions.entrySet()) {
+                buildColumns(entry.getKey(), entry.getValue());
+            }
+        }
+        catch (Exception e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, e.getMessage());
+        }
+        return tableDescriptions.get(new SchemaTableName(schemaName, tableName));
+    }
+
+    public ClusterSearchShardsResponse getSearchShards(ElasticsearchTableDescription tableDescription)
+    {
+        Client client = clients.get(tableDescription.getClusterName());
+        ClusterSearchShardsRequest request = new ClusterSearchShardsRequest(tableDescription.getIndex());
+        try {
+            return client.admin().cluster().searchShards(request).get();
+        }
+        catch (InterruptedException | ExecutionException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, e.getMessage());
+        }
+    }
+
+    private void buildColumns(SchemaTableName schemaTableName, ElasticsearchTableDescription tableDescription)
+            throws ExecutionException, InterruptedException, IOException, JSONException
```

Comment:
It's weird from an interface point of view that a method with name `buildColumns()` throws `ExecutionException, InterruptedException`, which are both low level stuff. I think this method itself should handle them.